### PR TITLE
feat(rest): Add filter functionality for GET /components route

### PR DESCRIPTION
**Pull request by blaumeiser-at-bosch**
feat(datahandler): Added filter functionality for resources

Signed-off-by: Henrik Engelbrink <henrik.engelbrink@bosch-si.com>

### CODE REVIEW

**Comment by imaykay** [comment]
I am going to do the code review.

File: libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGeneratorTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ResourcePredicateGeneratorTest {
```
Comment (imaykay) [inline libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGeneratorTest.java]: I would have liked another test checking the AND-functionality of the predicates. But as tests are rare in other places of the project, this should not block anything :-)

File: libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
```
Comment (imaykay) [inline libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGeneratorTest.java]: unused import...

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/ComponentController.java [hunk 3]
```diff
@@ -85,27 +80,25 @@
     @NonNull
     private final RestControllerHelper<Component> restControllerHelper;
 
+    @NonNull
+    private final ResourceListController<Component> resourceListController = new ResourceListController<>();
```
Comment (imaykay) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/ComponentController.java]: Again: why not managed?

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/ComponentController.java [hunk 4]
```diff
@@ -56,10 +54,7 @@
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Comment (imaykay) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/ComponentController.java]: Again unused imports - no new ones this time, but it should be easy to fix this when working on a class.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/ComponentController.java [hunk 5]
```diff
@@ -85,27 +80,25 @@
...
         User sw360User = restControllerHelper.getSw360UserFromAuthentication();
+        List<Component> allComponents = componentService.getComponentsForUser(sw360User);
 
-        List<Component> allComponents = new ArrayList<>();
-        if (name != null && !name.isEmpty()) {
-            allComponents.addAll(componentService.searchComponentByName(name));
-        } else {
-            allComponents.addAll(componentService.getComponentsForUser(sw360User));
-        }
+        requestParams = restControllerHelper.removePageableParamatersFromRequestParams(requestParams);
```
Comment (imaykay) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/ComponentController.java]: Why is this necessary? It might be easier if you could just tell me instead of me digging in the code :-)

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java [hunk 6]
```diff
@@ -67,6 +67,10 @@
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
 public class RestControllerHelper<T> {
 
+    public static final String REQUEST_PARAM_KEY_PAGE = "page";
```
Comment (imaykay) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java]: Currently there is no need for these fields to be public.

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGenerator.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.sw360.datahandler.resourcelists;
```
[the remark this answers is not in the document]
Comment (maxhbr) [reply libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGenerator.java]: I think the idea here was, to have the function generically available, so that it can also be used for the liferay UI or maybe even thrift

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGenerator.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        componentMap.put(Component._Fields.CREATED_ON, Component::getCreatedOn);
+        componentMap.put(Component._Fields.CREATED_BY, Component::getCreatedBy);
+        componentMap.put(Component._Fields.COMPONENT_TYPE, component -> Optional.ofNullable(component.getComponentType()).map(Object::toString).orElse(""));
+        componentMap.put(Component._Fields.COMPONENT_OWNER, Component::getComponentOwner);
+        componentMap.put(Component._Fields.OWNER_ACCOUNTING_UNIT, Component::getOwnerAccountingUnit);
+        componentMap.put(Component._Fields.OWNER_GROUP, Component::getOwnerGroup);
+        componentMap.put(Component._Fields.OWNER_COUNTRY, Component::getOwnerCountry);
+    }
+
+    <T extends TBase<?, ? extends TFieldIdEnum>> Predicate<T> predicateFromFilterMap(String resourceClass, Map<String, String> filter) {
```
[the remark this answers is not in the document]
Comment (maxhbr) [reply libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/resourcelists/ResourcePredicateGenerator.java]: the used `switch` does not work for `Class`. But I support your suggestion.

**Comment by maxhbr** [comment]
This PR will have conflicts with #447

**Comment by mcjaeger** [comment]
would it be Ok to close?

**Comment by mcjaeger** [comment]
as discussed, let's close it.
